nixos/doc: add md-to-db.sh, convert some sections to CommonMark, add 21.11 release notes stub

ryantm opened the pull request:
###### Motivation for this change

[RFC 72](https://github.com/NixOS/rfcs/blob/master/rfcs/0072-commonmark-docs.md)

See the added README and comments in md-to-db.sh for motivation too.

ryantm committed:
nixos/doc: add md-to-db.sh, convert "Building Your Own NixOS CD" to CommonMark

flokli commented on nixos/doc/manual/md-to-db.sh:
```diff
@@ -0,0 +1,32 @@
+
+# This script is temporarily needed while we transition the manual to
+# CommonMark. It converts the .md files in the regular manual folder
+# into DocBook files in the from_md folder.
```
I don't understand how this is used. This doesn't seem to be called anywhere in the build process.

Do we really want to keep the same manual pages in both docbook and md around in the repo at the same time, and ask everyone doing changes to manually keep them in sync?

Can't we consume `.md` files where things are already converted, and `.xml` where not, like in `doc/languages-frameworks`?

jtojnar replied:
The context is that Eelco [does not like Pandoc being in NixOS closure](https://github.com/NixOS/nixpkgs/pull/106847#issuecomment-744503875). But it has been in the closure for a long time now and I do not see much benefit trying to get rid of it during the migration period. Especially when the cost would be keeping artefacts in the repo.

flokli replied:
Keeping both files in sync won't work. People are not aware they need to edit both files when updating docs, and these files will just diverge.

Yes, pandoc is a heavy ask, but we also build half of [firefox due to polkit rules being in javascript](https://twitter.com/zimbatm/status/1341365282888146950), so having pandoc in the build closure is a small problem - and if people want to quickly iterate on a new rebuild, disabling the manual for that local test is probably the better way forward than dumping two different and slowly diverging documentation file formats.

jtojnar replied:
We could have tooling for that – I think adisbladis mentioned they use GitHub actions for that in one of their projects. But seriously, pandoc is less than 200 MB static binary and even the build closure (if someone cannot use the cached version) is [comparable to whatever will replace it](https://github.com/NixOS/nixpkgs/pull/106847#issuecomment-744475018). So removing it only makes sense if we intend to keep this dual arrangement after the migration as well.

ryantm replied:
I'll update this to add myself to CODEOWNERS so I can monitor/enforce proper changes to the files.

ryantm replied:
Oh, wait, so pandoc is already in the closure? Then I don't think we should do it this way.

edolstra replied:
> but we also build half of firefox due to polkit rules being in javascript, so having pandoc in the build closure is a small problem

Bloat in one area shouldn't be an excuse for adding more bloat.

I'm strongly :-1: (as in "veto") on having a GHC dependency in the NixOS base system, even if it's just a build-time dependency.

ryantm replied:
@jtojnar What is currently adding pandoc to the NixOS closure?

jtojnar replied:
Oh, I though this is for Nixpkgs manual. NixOS currently does not depend on Pandoc.

flokli replied:
What's the plan on rendering these NixOS docs in the base system after the migration to markdown is done? What's the closure size of that?

ryantm replied:
> What's the plan on rendering these NixOS docs in the base system after the migration to markdown is done? What's the closure size of that?

We do not have a plan.

ryantm replied:
> I'll update this to add myself to CODEOWNERS so I can monitor/enforce proper changes to the files.

I've added myself to CODEOWNERS

flokli replied:
> > What's the plan on rendering these NixOS docs in the base system after the migration to markdown is done? What's the closure size of that?
> 
> We do not have a plan.

Then we should probably re-think on this.
If there's no way to produce the manual on NixOS without getting pandoc into the (build) closure. we either need to live with that, or not have any docs in the NixOS base system (which also doesn't sound very good to me either).

@edolstra, what are your thoughts on this?

flokli replied:
Let's do that discussion in https://github.com/NixOS/nixpkgs/issues/63513 I think.

flokli commented:
I don't think this should be merged as-is until we have tooling to ensure programmatically (enforced Github Actions maybe?) there's no different versions of the two different versions.

Alternatively, if most of the migration is done by a script anyways, we could also have a long-living PR migrating ALL documentation at once, rebased every once and then, and on merge removing the docbook files it converted from.

zowoq commented on .github/workflows/nixos-manual.yml:
```diff
@@ -0,0 +1,19 @@
+  tests:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2.3.4
```
Why `@v2.3.4` and not just `@v2`?

ryantm replied:
I don't have a preference. Does v2 install the latest 2 major version?

ryantm commented:
> I don't think this should be merged as-is until we have tooling to ensure programmatically (enforced Github Actions maybe?) there's no different versions of the two different versions.

I've added a GitHub Action to this PR to enforce consistency in the generated Markdown files.

ryantm committed:
CODEOWNERS: add ryantm to /nixos/doc

jtojnar commented on nixos/doc/manual/from_md/README.md:
```diff
@@ -0,0 +1,5 @@
+This directory is temporarily needed while we transition the manual to CommonMark. It stores the output of the ../md-to-db.sh script that converts CommonMark files back to DocBook.
+
+We are choosing to convert the MarkDown to DocBook at authoring time instead of manual building time, because we do not want the pandoc toolchain to become part of the NixOS closure.
```
```suggestion
We are choosing to convert the Markdown to DocBook at authoring time instead of manual building time, because we do not want the pandoc toolchain to become part of the NixOS closure.
```

ryantm replied:
resolved

ryantm commented:
@floki maybe in light of the progress on #108063 you'd be okay with merging this so we can start working on converting the NixOS manual to CommonMark.

ryantm commented:
@jonringer and I talked and he 21.11 release notes to be in Markdown, so I think after the 21.05 branch off we should merge this and create those release notes in Markdown. I can try to update this so it adds the release note stub file too.

ryantm committed:
nixos/doc: convert "Contributing to this manual" to CommonMark

Also updates it to mention running md-to-db.sh.

ryantm committed:
nixos/doc: add 21.11 release notes stub

ryantm commented:
I've updated this to convert the "contributing to this manual" section (mentioning the need to run the `md-to-db.sh` script, and added the 21.11 release notes stub.

jonringer commented:
thanks for the release stub. Changes look great :)

jonringer commented:
The time has come!

zowoq commented on .github/workflows/nixos-manual.yml:
```diff
@@ -0,0 +1,20 @@
+name: NixOS manual checks
+
+on:
+  pull_request:
```
Just an FYI as actions have changed since the PR was opened: this won't run automatically for first time contributors so it's possible that the docs could drift out of sync.

https://github.blog/2021-04-22-github-actions-update-helping-maintainers-combat-bad-actors/

Would need to do the same as we did in https://github.com/NixOS/nixpkgs/commit/5ba2e4e9c65d32ea3d7def475655227a3835d2dc (`pull_request_target`, `permissions:`, and the PR merge ref) to make it run unconditionally.

jonringer replied:
ah, great point.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/happy-docsgiving/10214/27
